src/main/scala/mesosphere/mesos/TaskBuilder.scala

```diff
@@ -318,10 +321,38 @@ object TaskBuilder {
         "MESOS_TASK_ID" -> taskId.map(_.getValue),
         "MARATHON_APP_ID" -> Some(app.id.toString),
         "MARATHON_APP_VERSION" -> Some(app.version.toString),
-        "MARATHON_APP_DOCKER_IMAGE" -> app.container.flatMap(_.docker.map(_.image))
+        "MARATHON_APP_DOCKER_IMAGE" -> app.container.flatMap(_.docker.map(_.image)),
+        "MARATHON_RESOURCE_CPUS" -> Some(app.cpus.toString),
+        "MARATHON_RESOURCE_MEM" -> Some(app.mem.toString),
+        "MARATHON_RESOURCE_DISK" -> Some(app.disk.toString)
       ).collect {
           case (key, Some(value)) => key -> value
-        }.toMap
+        }.toMap ++ labelsToEnvVars(app.labels)
+    }
+  }
+
+  def labelsToEnvVars(labels: Map[String, String]): Map[String, String] = {
+    var variables = Map(
+      "MARATHON_APP_LABELS" -> labels.keys.map(escapeEnvironmentVariableName).mkString(" ")
+    )
+
+    labels.foreach {
+      case (key, value) => {
+        variables = variables ++ Map(s"${labelEnvironmentKeyPrefix}${escapeEnvironmentVariableName(key)}" -> value)
+      }
+    }
+
+    variables
+  }
+
+  def escapeEnvironmentVariableName(name: String): String = {
+    val variableName = name.replaceAll("[^a-zA-Z0-9_]+", "_").toUpperCase
```

Comment:
This would replace subsequent invalid chars into a single underscore: e.g. `1234$$$$` into `1234_`. If you omit the leading `+`, then the string would be replaced to `1234____`
I am not exactly sure which one is better...